### DISCUSSION TITLE
Look-up based objective function

### DISCUSSION THREAD

**Discussion by sadvaseb** [cvxpy/cvxpy#1832] [Q&A]
Hi all,

I use cvxpy for an allocation problem. Like most allocation problems, the objective function is not a continuous mathematical function. The objective value is captured from a table that contains the objective value for each action(e.g. if you pick action X, its objection value is Y). So, the objective function has to match the action (or a set of calculations based on the action) with the value from the "look-up table". However, it seems cvxpy does not support look-up, indexing, merge, or similar matrix operations. 

In my case, we have an integer programming problem (binary) where we can pick one or more actions from a set of actions. Our cvxpy.Variable is:

`element_vars = [1, 0, 0, 1] (p.s. element_vars is a Cvxpy placeholder with no value, but I put some numbers here for demonstration.)`

And the set of actions to pick is:

`actions = [2, 3, 4, 5]`

The objective function depends based on "total actions" (not single actions) which is calculated as:

`total_actions = cvxpy.matmul(element_vars, actions.T) = [2+5] = [7]`

The objective value is picked from a look-up table based on "total_actions" value. For example, if the look-up table is:

```
lookup_table = [[2, 3, 4, 5, 6, 7, 8, 9],
                [1, 2, 6, 8, 3, 4, 8,  9]]
```

Where the first row shows "total_actions" and the second row shows its corresponding objective value. In this example, the objective value for total_actions=7 is 4.

Since element_vars is a placeholder, and it only exists inside the CVXPy environment, I can not use Pandas, df, or NumPy functions. I tried the below code

```
total_actions = cvxpy.matmul(actions, element_vars)
self.objective_terms.append(cvxpy.multiply(lookup_table[np.where(lookup_table[:,0] == total_actions),1], weight))
```
But, I get the below error

`Error in python process: At line 23: <class 'Exception'>: Cannot evaluate the truth value of a constraint or chain constraints, e.g., 1 >= x >= 0.`
I also tried to pass the lookup_table as a Cvxpy parameter, but no progress. 

Here are my questions:

1. Is there any function to process an array (i.e., look for a value inside an array and return the value or index) in the CVXPy environment?
2. Does it worth building look-up based objective function for cvxpy, considering its wide use? 

Thanks in advance!

**Comment by SteveDiamond** [comment]
Thanks for posting this interesting use case! You cannot index cvxpy expressions using other cvxpy expressions. You also cannot apply ``np.where`` to a cvxpy expression. So the answer to question 1 is only normal indexing or indexing with numpy arrays is supported.

In terms of question 2, this has come up before but is getting further into MIP world than we usually go. However, it's certainly worth discussing and I'll bring it up on the discord (https://discord.gg/4urRQeGBCr).

**Reply by sadvaseb** [reply]
Thanks for taking the time to answer my questions !
When you say "normal indexing or indexing with numpy arrays", does that mean we can use list comprehension?
I tried to use basic list comprehension for this:

```
lookup_actions = list(lookup_table[0,:])    # action list
lookup_value = list(lookup_table[1,:])    # value list

self.objective_terms.append(cp.multiply(
              [lookup_value[i] if total_actions == lookup_actions[i] else 0 for i in range(len(lookup_value ))]
                                                    , weight))
```
It seems cvxpy is not able to do any compassion (i.e., "==") between a number and a placeholder variable. I get the same error. Do you have any suggestions on how to code this specific scenario?

Thanks again for your kind help!

**Reply by SteveDiamond** [reply]
Can you try using a piecewise constant objective function as described here: https://github.com/cvxpy/cvxpy/issues/1144
Sorry I wasn't more clear. Your approach involving indexing into something using cvxpy expressions is fundamentally not possible. By "normal indexing" I meant like ``x[:2]``.
